[CL-29] Modal Components for the component library

Thomas-Avery opened the pull request:
## Type of change

<!-- (mark with an `X`) -->

```
- [ ] Bug fix
- [X] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

<!--Describe what the purpose of this PR is. For example: what bug you're fixing or what new feature you're adding-->

The purpose of this PR is to add Modals into the component library.

To note, this PR is focused on the visual of the modal that will "pop-up" and not the actual pop-up implementation.

The two modal templates are as follows:

### Simple Modal

### Default Icon
```
<bit-simple-modal>
  <span bit-modal-title>Alert Modal Title</span>
  <span bit-modal-content>Message Content</span>
   <div bit-modal-footer class="tw-flex tw-flex-row tw-gap-2">
      <button bitButton buttonType="primary"> Yes </button>
      <button bitButton buttonType="secondary"> No </button>
   </div>
</bit-simple-modal>
```

### Custom Icon
```
<bit-simple-modal>
  <i bit-modal-icon class="bwi bwi-star tw-text-3xl tw-text-success" aria-hidden="true"></i>
  <span bit-modal-title>Premium Subscription Available</span>
  <span bit-modal-content>Message Content</span>
   <div bit-modal-footer class="tw-flex tw-flex-row tw-gap-2">
      <button bitButton buttonType="primary"> Yes </button>
      <button bitButton buttonType="secondary"> No </button>
   </div>
</bit-simple-modal>
```

The keywords bit-modal-icon, bit-modal-title, bit-modal-content, and bit-modal-footer determine placement of content in the modal using angular's multi-slot content project.
https://angular.io/guide/content-projection#multi-slot-content-projection

### Modal
```
 <bit-modal [modalSize]="modalSize">
    <span bit-modal-title> Modal Title </span>
    <span bit-modal-content>
    Modal body text goes here.
    </span>
    <div bit-modal-footer class="tw-flex tw-flex-row tw-gap-2">
      <button bitButton buttonType="primary"> Save </button>
      <button bitButton buttonType="secondary"> Cancel </button>
    </div>
  </bit-modal>
```
Where "modalSize" can be small, default, or large, controlling the width of the modal displayed.

The keywords bit-modal-title, bit-modal-content, and bit-modal-footer determine placement of content in the modal using angular's multi-slot content project.
https://angular.io/guide/content-projection#multi-slot-content-projection

## Code changes

<!--Explain the changes you've made to each file or major component. This should help the reviewer understand your changes-->
<!--Also refer to any related changes or PRs in other repositories-->

Code for the simple modal template.

libs/components/src/modal/modal-simple.component.html
libs/components/src/modal/modal-simple.component.ts
libs/components/src/modal/modal-simple.stories.ts

Code for the modal template.
libs/components/src/modal/modal.component.html
libs/components/src/modal/modal.component.ts
libs/components/src/modal/modal.stories.ts

The Modals components can be viewed by running `npm run storybook` at the root of the project.


## Screenshots

### Simple Modal
![image](https://user-images.githubusercontent.com/43214426/178328224-ea19dfbc-9af2-4338-bc6f-b74eaadf6750.png)

![image](https://user-images.githubusercontent.com/43214426/178328307-33f24a6d-0125-4390-93e5-49c4c2657beb.png)


### Modal

![image](https://user-images.githubusercontent.com/43214426/178328936-9fe2573d-bd17-493c-8457-f9d525398f88.png)

![image](https://user-images.githubusercontent.com/43214426/178329107-00839815-90ab-4833-8668-5117b831c203.png)


![image](https://user-images.githubusercontent.com/43214426/178328986-470e33d8-762d-478d-832c-c9d875ddb578.png)


## Before you submit

<!-- (mark with an `X`) -->

```
- [X] I have checked for **linting** errors (`npm run lint`) (required)
- [ ] I have added **unit tests** where it makes sense to do so (encouraged but not required)
- [ ] This change requires a **documentation update** (notify the documentation team)
- [ ] This change has particular **deployment requirements** (notify the DevOps team)
```

Hinton reviewed (commented):
Some general feedback.

1. We should avoid non standard size. Check with design which [sizes](https://tailwindcss.com/docs/max-width) they would like to use. We should probably also use `max-width` instead of `width` to be a bit responsive.
2. Try and use a single width per modal size, the body and title widths can be generated using margins and/or percentages.

As to the height, https://tailwindcss.com/docs/max-height combined with padding should work? Might need to put the max-height on a parent container, and use height 100% on the inner though.

Hinton commented:
@danielleflinn for the simple modals, do we want the developers to use any icons or just a set number, similar to callouts?

danielleflinn commented:
@Hinton lets default the component to the bwi-exclamation-triangle, but allow it to be configurable to any icon per implementation.

Thomas-Avery commented:
Visually, everything should now be implemented and is ready for review. 

- fixed height scrolling to work as requested
- responsive width now implemented
- added the "X" close icon to Modals
- moved all examples to using bitButtons 
- added example stories to show content expansion and scrolling for both Simple Modal and Modal

Here is the proposal for implementing the default bwi-exclamation-triangle icon. 
This proposal is now implemented as well.

## Simple Modal

### Default Icon
```
<bit-simple-modal [useDefaultIcon]="true">
  <span title>Alert Modal Title</span>
  <span message>Message Content</span>
  <button footer bitButton buttonType="primary">Yes</button>
  <button footer bitButton buttonType="secondary">No</button>
</bit-simple-modal>
```

### Custom Icon
```
<bit-simple-modal [useDefaultIcon]="false">
  <i icon class="bwi bwi-star tw-text-3xl tw-text-success" aria-hidden="true"></i>
  <span title>Premium Subscription Available</span>
  <span message>Message Content</span>
  <button footer bitButton buttonType="primary">Yes</button>
  <button footer bitButton buttonType="secondary">No</button>
</bit-simple-modal>
```

Instead of using the keyword vector, I think "icon" works better for the template.

Hinton commented on libs/components/src/modal/modal-simple.stories.ts:
```diff
@@ -0,0 +1,180 @@
+      </span>
+      <span message> Message Content
+      Message text goes here.<br>
+      repeating lines of characters
```
Lets make this an `ng-for`, so we can actually follow what happens here.

Thomas-Avery replied:
solved in https://github.com/bitwarden/clients/pull/3053/commits/dd115078508a8bf1c648d415eeb0baafbd2ceab8

Hinton commented on libs/components/src/modal/modal-simple.component.ts:
```diff
@@ -0,0 +1,10 @@
+  templateUrl: "./modal-simple.component.html",
+})
+export class ModalSimpleComponent {
+  @Input()
+  useDefaultIcon: boolean;
```
We should be able to use `ViewChild` here to determine if the icon exists.

Thomas-Avery replied:
resolved in https://github.com/bitwarden/clients/pull/3053/commits/27fbba52ff92a089fbd15c77a43ee7757c434671

Hinton reviewed (commented):
Some initial review comments while I review the rest.

Hinton commented on libs/components/src/modal/modal.component.ts:
```diff
@@ -0,0 +1,25 @@
+})
+export class ModalComponent {
+  // Controls the size of the bit modal displayed. Needs to be implemented.
+  @Input()
+  modalSize: "small" | "default" | "large";
```
Lets keep these on the same line.

```suggestion
  @Input() modalSize: "small" | "default" | "large";
```

Hinton commented on libs/components/src/modal/modal.component.ts:
```diff
@@ -0,0 +1,25 @@
+  @Input()
+  modalSize: "small" | "default" | "large";
+
+  getWidth() {
```
Let's change this to a getter.

Thomas-Avery replied:
resolved in https://github.com/bitwarden/clients/pull/3053/commits/dd115078508a8bf1c648d415eeb0baafbd2ceab8

Hinton commented on libs/components/src/modal/modal.component.html:
```diff
@@ -0,0 +1,51 @@
+    <div
+      class="tw-w-full tw-flex tw-flex-row tw-items-center tw-p-4 tw-gap-2 tw-max-h-full tw-flex-none tw-order-1 tw-self-stretch tw-grow-0"
+    >
+      <ng-content select="[footer]"></ng-content>
```
Let's use the `bit` prefix on these selectors.

Thomas-Avery replied:
Should we go with:

bit-modal-title
bit-modal-body
bit-modal-footer

or 

bit-title
bit-body
bit-footer

The reason I bring it up is if there will ever be any future conflicts with other bit-<something> keywords.

Hinton commented on libs/components/src/modal/modal.stories.ts:
```diff
@@ -0,0 +1,178 @@
+
+import { ModalComponent } from "./modal.component";
+
+export default {
+  title: "Component Library/Modals/Modal",
+  component: ModalComponent,
+  decorators: [
+    moduleMetadata({
+      imports: [ButtonModule],
+    }),
+  ],
+  args: {
+    modalSize: "small",
+  },
+} as Meta;
```
Can we add the design link here?

Thomas-Avery replied:
resovled in https://github.com/bitwarden/clients/pull/3053/commits/dd115078508a8bf1c648d415eeb0baafbd2ceab8

Hinton commented on libs/components/src/modal/modal.component.html:
```diff
@@ -0,0 +1,51 @@
+  </div>
+
+  <!-- Modal Footer-->
+  <div
+    class="tw-w-full tw-flex tw-flex-col tw-justify-end tw-items-end tw-p-0 tw-max-h-full tw-bg-background-alt tw-flex-none tw-order-2 tw-self-stretch tw-grow-0"
+  >
+    <!-- Modal Footer Ruler-->
+    <div
+      class="tw-w-full tw-box-border tw-h-px tw-bg-secondary-300 tw-border-solid tw-border-secondary-300 tw-border-0 tw-border-t tw-flex-none tw-order-none tw-self-stretch tw-grow-0"
+    ></div>
+    <!-- Modal Footer Buttons-->
+    <div
+      class="tw-w-full tw-flex tw-flex-row tw-items-center tw-p-4 tw-gap-2 tw-max-h-full tw-flex-none tw-order-1 tw-self-stretch tw-grow-0"
+    >
```
Why do we need 3 classes for the footer? Could we not use a single div for everything?

Thomas-Avery replied:
We can merge the footer ruler into the main footer div. 

The modal footer buttons div provides the padding and gap to center the buttons inside the main footer div. 

![image](https://user-images.githubusercontent.com/43214426/178599229-624403ae-9a10-4176-ac06-516e30f61ae5.png)

Unless I'm missing something, are you wanting consumers to implement their own padding and gap styling?

I just was going off the figma models, which have 3 separate containers.

Hinton replied:
I'm not really following. What's stopping us from having padding on the outer div?

Thomas-Avery replied:
resolved in https://github.com/bitwarden/clients/pull/3053/commits/c4d09a5eaa0ff6d0cfc5a943fe8938a48a03b039

Hinton commented on libs/components/src/modal/modal.component.html:
```diff
@@ -0,0 +1,51 @@
+  </div>
+
+  <!-- Modal Body-->
+  <div
+    class="tw-w-full tw-flex tw-flex-col tw-justify-center tw-items-center tw-order-1 tw-self-stretch tw-grow-0 tw-relative tw-flex-auto tw-overflow-y-auto"
+  >
+    <!-- Modal Body Text-->
+    <div
+      class="tw-w-full tw-pb-8 tw-p-4 tw-not-italic tw-font-normal tw-text-base tw-text-main tw-order-none tw-self-stretch tw-grow-0 tw-relative tw-flex-auto tw-overflow-y-auto"
+    >
+      <ng-content select="[body]"></ng-content>
+    </div>
+  </div>
```
Do we need two divs here?

Thomas-Avery replied:
Same as the footer, the body text div provides padding to center inside the main body container.

Thomas-Avery replied:
resolved in https://github.com/bitwarden/clients/pull/3053/commits/c4d09a5eaa0ff6d0cfc5a943fe8938a48a03b039

Hinton commented on libs/components/src/modal/modal.component.html:
```diff
@@ -0,0 +1,51 @@
+>
+  <!-- Modal Header-->
+  <div
+    class="tw-w-full tw-box-border tw-flex tw-flex-row tw-items-center tw-gap-4 tw-p-4 tw-max-h-full tw-border-0 tw-border-b tw-border-solid tw-border-secondary-300 tw-flex-none tw-order-none tw-self-stretch tw-grow-0"
```
Do you need `order-none`?

Thomas-Avery replied:
resolved in https://github.com/bitwarden/clients/pull/3053/commits/c4d09a5eaa0ff6d0cfc5a943fe8938a48a03b039

Thomas-Avery commented on libs/components/src/modal/modal.component.ts:
```diff
@@ -0,0 +1,25 @@
+  templateUrl: "./modal.component.html",
+})
+export class ModalComponent {
+  // Controls the size of the bit modal displayed. Needs to be implemented.
```
resolved in https://github.com/bitwarden/clients/pull/3053/commits/dd115078508a8bf1c648d415eeb0baafbd2ceab8

Thomas-Avery commented on libs/components/src/modal/modal.component.html:
```diff
@@ -0,0 +1,51 @@
+<!-- Popup shell -->
+<div
+  [ngClass]="getWidth()"
+  class="tw-max-h-screen tw-my-4 tw-box-border tw-flex tw-flex-col tw-items-start tw-p-0 tw-bg-text-contrast tw-border tw-border-solid tw-border-secondary-300 tw-rounded tw-flex-none tw-order-0 tw-grow-1"
```
resolved in https://github.com/bitwarden/clients/pull/3053/commits/a8ea95db1f9b8c0a4e729c3893e6740a627a47b0

Hinton commented on libs/components/src/modal/modal.component.html:
```diff
@@ -0,0 +1,35 @@
+
+  <!-- Modal Footer-->
+  <div
+    class="tw-w-full tw-flex tw-flex-row tw-items-center tw-p-4 tw-gap-2 tw-max-h-full tw-bg-background-alt tw-flex-none tw-order-2 tw-self-stretch tw-grow-0 tw-border-solid tw-border-secondary-300 tw-border-0 tw-border-t"
```
```suggestion
    class="tw-p-4 tw-bg-background-alt tw-border-solid tw-border-secondary-300 tw-border-0 tw-border-t"
```

Thomas-Avery replied:
resolved in https://github.com/bitwarden/clients/pull/3053/commits/9f937eaff562bd6e7f050066a3448336049c9436

Hinton commented on libs/components/src/modal/modal.component.html:
```diff
@@ -0,0 +1,35 @@
+  </div>
+
+  <!-- Modal Body-->
+  <div
+    class="tw-pb-8 tw-p-4 tw-w-full tw-flex tw-flex-col tw-order-1 tw-self-stretch tw-grow-0 tw-relative tw-flex-auto tw-overflow-y-auto tw-not-italic tw-font-normal tw-text-base tw-text-main"
+  >
```
```suggestion
  <div class="tw-pb-8 tw-p-4 tw-w-full tw-relative tw-overflow-y-auto">
```

Thomas-Avery replied:
resolved in https://github.com/bitwarden/clients/pull/3053/commits/9f937eaff562bd6e7f050066a3448336049c9436

Hinton commented on libs/components/src/modal/modal.component.html:
```diff
@@ -0,0 +1,35 @@
+      <ng-content select="[bit-modal-title]"></ng-content>
+    </h2>
+    <!-- Modal Title close button container TODO: when implementing modal popup this should broadcast to close modal popup-->
+    <button class="tw-border-0 tw-bg-transparent tw-text-main tw-p-0">
```
```suggestion
    <button class="tw-border-0 tw-bg-transparent tw-p-0">
```

Thomas-Avery replied:
resolved in https://github.com/bitwarden/clients/pull/3053/commits/9f937eaff562bd6e7f050066a3448336049c9436

Hinton commented on libs/components/src/modal/modal.component.html:
```diff
@@ -0,0 +1,35 @@
+    </h2>
+    <!-- Modal Title close button container TODO: when implementing modal popup this should broadcast to close modal popup-->
+    <button class="tw-border-0 tw-bg-transparent tw-text-main tw-p-0">
+      <i class="bwi bwi-close tw-text-xs tw-font-bold tw-text-main" aria-hidden="true"></i>
```
```suggestion
      <i class="bwi bwi-close tw-text-xs tw-font-bold" aria-hidden="true"></i>
```

Thomas-Avery replied:
resolved in https://github.com/bitwarden/clients/pull/3053/commits/9f937eaff562bd6e7f050066a3448336049c9436

Hinton replied:
Hmm that's true, bootstrap styling messing things up again.

Hinton reviewed (commented):
There are quite a few unnecessary css classes we can remove, I've posted my suggestions for the regular modal. Please look into doing the same for the simple modal, as the fewer classes we use the easier it is to maintain.

We should also remove the html comments unless they bring value, since we've eliminated most of the unnecessary divs I don't view the comments as providing much value any longer as the selector already describes what's in the divs.

Thomas-Avery commented on libs/components/src/modal/modal.component.html:
```diff
@@ -0,0 +1,35 @@
+  >
+    <!-- Modal Title text container -->
+    <h2
+      class="tw-text-lg tw-font-normal tw-leading-7 tw-uppercase tw-flex tw-items-center tw-text-main tw-flex-none tw-order-none tw-grow tw-mb-0"
```
resolved in https://github.com/bitwarden/clients/pull/3053/commits/9f937eaff562bd6e7f050066a3448336049c9436

Thomas-Avery commented on libs/components/src/modal/modal.component.html:
```diff
@@ -0,0 +1,35 @@
+>
+  <!-- Modal Header-->
+  <div
+    class="tw-w-full tw-flex tw-flex-row tw-items-center tw-gap-4 tw-p-4 tw-max-h-full tw-border-0 tw-border-b tw-border-solid tw-border-secondary-300 tw-flex-none tw-self-stretch tw-grow-0"
```
Resolved in https://github.com/bitwarden/clients/pull/3053/commits/9f937eaff562bd6e7f050066a3448336049c9436

Thomas-Avery commented on libs/components/src/modal/modal.component.html:
```diff
@@ -0,0 +1,35 @@
+<!-- Popup shell -->
+<div
+  [ngClass]="width"
+  class="tw-max-h-screen tw-my-4 tw-flex tw-flex-col tw-items-start tw-p-0 tw-bg-text-contrast tw-border tw-border-solid tw-border-secondary-300 tw-rounded tw-overflow-hidden"
```
resolved in https://github.com/bitwarden/clients/pull/3053/commits/9f937eaff562bd6e7f050066a3448336049c9436

did not take out tw-overflow-hidden as discussed.

Hinton commented on libs/components/src/modal/modal.component.html:
```diff
@@ -0,0 +1,25 @@
+  class="tw-max-h-screen tw-my-4 tw-flex tw-flex-col tw-bg-text-contrast tw-border tw-border-solid tw-border-secondary-300 tw-rounded tw-text-main tw-overflow-hidden"
+>
+  <div
+    class="tw-flex tw-gap-4 tw-p-4 tw-border-0 tw-border-b tw-border-solid tw-border-secondary-300 tw-text-main"
```
Color gets inherited from the parent div.
```suggestion
    class="tw-flex tw-gap-4 tw-p-4 tw-border-0 tw-border-b tw-border-solid tw-border-secondary-300"
```

Thomas-Avery replied:
resolved in https://github.com/bitwarden/clients/pull/3053/commits/938f2362c707729d900ffabad29b2289efe7fd5a

Hinton commented on libs/components/src/modal/modal.component.html:
```diff
@@ -0,0 +1,25 @@
+  <div
+    class="tw-flex tw-gap-4 tw-p-4 tw-border-0 tw-border-b tw-border-solid tw-border-secondary-300 tw-text-main"
+  >
+    <h2 class="tw-text-lg tw-uppercase tw-flex tw-grow tw-mb-0">
```
Is display flex needed here?
```suggestion
    <h2 class="tw-text-lg tw-uppercase tw-grow tw-mb-0">
```

Thomas-Avery replied:
resolved in https://github.com/bitwarden/clients/pull/3053/commits/938f2362c707729d900ffabad29b2289efe7fd5a

Hinton commented on libs/components/src/modal/modal.component.html:
```diff
@@ -0,0 +1,25 @@
+    </button>
+  </div>
+
+  <div class="tw-pb-8 tw-p-4 tw-w-full tw-relative tw-overflow-y-auto">
```
Is Width and relative needed here?

```suggestion
  <div class="tw-pb-8 tw-p-4 tw-overflow-y-auto">
```

Thomas-Avery replied:
resolved in https://github.com/bitwarden/clients/pull/3053/commits/938f2362c707729d900ffabad29b2289efe7fd5a

Hinton commented on libs/components/src/modal/modal-simple.component.html:
```diff
@@ -0,0 +1,23 @@
+<div
+  class="tw-max-w-sm tw-max-h-screen tw-my-4 tw-flex tw-flex-col tw-bg-text-contrast tw-border tw-border-solid tw-border-secondary-300 tw-rounded tw-text-main tw-overflow-hidden"
+>
+  <div
+    class="tw-text-center tw-text-main tw-flex tw-flex-col tw-items-center tw-px-4 tw-pt-4 tw-gap-2"
+  >
```
```suggestion
  <div class="tw-text-center tw-flex tw-flex-col tw-items-center tw-px-4 tw-pt-4 tw-gap-2">
```

Thomas-Avery replied:
resolved in https://github.com/bitwarden/clients/pull/3053/commits/938f2362c707729d900ffabad29b2289efe7fd5a

Hinton commented on libs/components/src/modal/modal-simple.component.html:
```diff
@@ -0,0 +1,23 @@
+    <ng-template #elseBlock>
+      <i class="bwi bwi-exclamation-triangle tw-text-3xl tw-text-warning" aria-hidden="true"></i>
+    </ng-template>
+    <h2 class="tw-font-semibold tw-text-base tw-flex-none tw-mb-0">
```
```suggestion
    <h2 class="tw-font-semibold tw-text-base tw-mb-0">
```

Thomas-Avery replied:
resolved in https://github.com/bitwarden/clients/pull/3053/commits/938f2362c707729d900ffabad29b2289efe7fd5a

Hinton commented on libs/components/src/modal/modal-simple.component.html:
```diff
@@ -0,0 +1,23 @@
+      <ng-content select="[bit-sm-title]"></ng-content>
+    </h2>
+  </div>
+  <div
+    class="tw-text-center tw-text-main tw-text-base tw-pt-2 tw-pb-4 tw-px-4 tw-relative tw-overflow-y-auto"
+  >
```
```suggestion
  <div class="tw-text-center tw-text-base tw-pt-2 tw-pb-4 tw-px-4 tw-overflow-y-auto">
```

Thomas-Avery replied:
resolved in https://github.com/bitwarden/clients/pull/3053/commits/938f2362c707729d900ffabad29b2289efe7fd5a

Hinton commented on libs/components/src/modal/modal-simple.stories.ts:
```diff
@@ -0,0 +1,85 @@
+  props: args,
+  template: `
+  <bit-simple-modal>
+      <span bit-sm-title> Alert Modal
```
Lets use the same name for these across both components, so `bit-modal-title`.

Thomas-Avery replied:
Agreed: 

These names I think fit both well:

`bit-modal-title`
`bit-modal-content`
`bit-modal-footer`

and for the simple modal as well `bit-modal-icon`

Hinton reviewed (commented):
I think everything looks good from a code perspective. @danielleflinn do you mind reviewing the design before we merge this?
